torch/_jit_internal.py

```diff
@@ -421,18 +445,36 @@ def fn(*args, **kwargs):
         else:
             return if_false(*args, **kwargs)
 
-    if if_true.__doc__ is None and if_false.__doc__ is not None:
-        doc = if_false.__doc__
-        if_true.__doc__ = doc
-    elif if_false.__doc__ is None and if_true.__doc__ is not None:
-        doc = if_true.__doc__
-        if_false.__doc__ = doc
-    elif if_false.__doc__ is None and if_true.__doc__ is None:
-        # neither function has a docstring
-        doc = None
+    if_true_signature = inspect.signature(if_true)
+    if_false_signature = inspect.signature(if_false)
+    if if_true_signature.parameters == if_false_signature.parameters:
+        parameters = list(if_true_signature.parameters.values())
     else:
-        raise RuntimeError("only one function can have a docstring")
-    fn.__doc__ = doc
+        if len(if_true_signature.parameters) != len(if_false_signature.parameters):
+            raise RuntimeError(
+                f"The number of parameters in signature of {getattr(if_true, '__name__', repr(if_true))} and "
+                f"{getattr(if_false, '__name__', repr(if_false))} mismatch: "
+                f"{len(if_true_signature.parameters)} != {len(if_false_signature.parameters)}."
+            )
+
+        parameters = [
+            inspect.Parameter(
+                name=extract_attribute(if_true_parameter, if_false_parameter, "name"),
+                kind=extract_attribute(if_true_parameter, if_false_parameter, "kind"),
+                default=extract_attribute(
+                    if_true_parameter, if_false_parameter, "default", sentinel=inspect.Parameter.empty
+                ),
+                annotation=extract_attribute(
+                    if_true_parameter, if_false_parameter, "annotation", sentinel=inspect.Parameter.empty
+                ),
+            )
+            for if_true_parameter, if_false_parameter in zip(
+                if_true_signature.parameters.values(), if_false_signature.parameters.values()
+            )
+        ]
+    fn.__signature__ = inspect.Signature(parameters)
```

Comment:
This only sets the parameters, but leaves the `return_annotation`. The problem is that some dispatched operators have different return types, for example `torch.unique`.

This makes no difference for the documentation, since we don't show the annotation anyway. But even if the `return_annotation` is the same, it will be missing if someone `inspect`'s the signature of the returned `fn`.